esp/services/ws_dfu/ws_dfuService.cpp

```diff
@@ -1543,6 +1549,99 @@ bool CWsDfuEx::onDFUArrayAction(IEspContext &context, IEspDFUArrayActionRequest
+        context.ensureSuperUser(ECLWATCH_SUPER_USER_ACCESS_DENIED, "Access denied, administrators only.");
+#endif
+
+    if (isDetachedFromDali())
+        throw makeStringException(ECLWATCH_INVALID_INPUT, "ESP server is detached from Dali. Please try later.");
+
+    StringBuffer userID;
+    context.getUserID(userID);
+    Owned<IUserDescriptor> userDesc;
+    if (userID.isEmpty())
+        userID.set("hpcc"); //The userID is needed for setProtect().
+    else
+    {
+        userDesc.setown(createUserDescriptor());
+        userDesc->set(userID, context.queryPassword(), context.querySignature());
+    }
+
+    for (unsigned i = 0; i < req.getLogicalFiles().length(); i++)
+    {
+        const char *file = req.getLogicalFiles().item(i);
+        if (isEmptyString(file))
+            continue;
+
+        Owned<IDistributedFile> df = queryDistributedFileDirectory().lookup(file, userDesc, false, false, true, nullptr, defaultPrivilegedUser); // lock super-owners
+        if (!df)
+            throw MakeStringException(ECLWATCH_FILE_NOT_EXIST, "Cannot find file %s.", file);
+
+        if (protect != CDFUChangeProtection_UnprotectAll)
+            df->setProtect(userID, protect == CDFUChangeProtection_Protect ? true : false);
+        else
+            clearFileProtections(df);
+    }
+    return true;
+}
+
+void CWsDfuEx::clearFileProtections(IDistributedFile *df)
+{
+    StringArray owners;
+    Owned<IPropertyTreeIterator> itr = df->queryAttributes().getElements("Protect");
+    ForEach(*itr)
+    {
+        const char *owner = itr->query().queryProp("@name");
+        if (!isEmptyString(owner))
+            owners.append(owner);
+    }
+
+    ForEachItemIn(i, owners)
+        df->setProtect(owners.item(i), false);
+}
+
+bool CWsDfuEx::changeFileRestrictions(IEspContext &context, IEspDFUArrayActionRequest &req, IEspDFUArrayActionResponse &resp)
+{
+    CDFUChangeRestriction changeRestriction = req.getRestrict();
+    if (changeRestriction == CDFUChangeRestriction_NoChange)
+        return true;
+
+    context.ensureFeatureAccess("DFURestrictedAccess", SecAccess_Write, ECLWATCH_DFU_ACCESS_DENIED, "DFURestrictedAccess: Permission denied.");
+
+    if (isDetachedFromDali())
+        throw makeStringException(ECLWATCH_INVALID_INPUT, "ESP server is detached from Dali. Please try later.");
+
+    StringBuffer userID;
+    context.getUserID(userID);
+    Owned<IUserDescriptor> userDesc;
+    if (!userID.isEmpty())
+    {
+        userDesc.setown(createUserDescriptor());
+        userDesc->set(userID, context.queryPassword(), context.querySignature());
+    }
+
+    for (unsigned i = 0; i < req.getLogicalFiles().length();i++)
+    {
+        const char *file = req.getLogicalFiles().item(i);
+        if (isEmptyString(file))
+            continue;
+
+        Owned<IDistributedFile> df = queryDistributedFileDirectory().lookup(file, userDesc, false, false, true, nullptr, defaultPrivilegedUser); // lock super-owners
```

Comment:
Seems like you are changing some in the list and then if one isn't found you quit.  Should you gather exceptions and keep trying the rest of the list?

Reply:
Good point. Will do.